prefect_utils/hubspot_leads.py

```diff
@@ -0,0 +1,106 @@
+from collections import namedtuple
+
+import backoff
+import prefect
+import requests
+from prefect import task
+from snowflake.connector import ProgrammingError
+
+from . import snowflake
+
+# masters_leads is an iterable of strings like "Master's in Data Science"
+LeadsData = namedtuple('LeadsData', ['lms_user_id', 'masters_leads'])
+
+# We can only update 75 attributes at a time:
+# https://www.braze.com/docs/api/endpoints/user_data/post_user_track/
+ATTRIBUTE_BATCH_SIZE = 75
+
+# Attribute arrays can only hold 25 items, though we could ask for this limit to be bumped. As of the time of this
+# writing, our longest leads list is 14 items long (and only 4 users have more than 10). I think we're OK for now.
+# https://www.braze.com/docs/api/objects_filters/user_attributes_object/
+MASTERS_LEADS_LIMIT = 25
+
+
+@task
+@backoff.on_exception(backoff.expo,
+                      ProgrammingError,
+                      max_tries=3)
+def sync_hubspot_leads_to_braze(
+    sf_credentials: snowflake.SFCredentials,
+    sf_role: str,
+    sf_database: str,
+    sf_schema: str,
+    braze_api_key: str,
+    braze_api_server: str,
+):
+    logger = prefect.context.get("logger")
+
+    sf_connection = snowflake.create_snowflake_connection(
+        sf_credentials,
+        sf_role,
+    )
+    sf_table = 'hubspot_leads'
+    query = """
+    SELECT lms_user_id, masters_leads FROM {table} WHERE lms_user_id IS NOT NULL
```

Comment:
Do we want to send the entirety of this table each time or just new additions for each run? I suppose the list of masters programs that a user can be a lead for can change over time

Reply:
Yes, I believe leads can change over time.

Do we have existing best-practices for only updating changes? Something like "(1) record last-updated time in data mart (2) record the last time the task and/or flow ran (3) use that info in the SELECT"

Reply:
I think I my question was more concerned with braze would handle this sort of thing. Like would we be constantly appending potentially redundant data or would it overwrite what we had previously sent to that API? I tend to think that since leads can change overtime, it would be best if it did overwrite the previous data in Braze.

To answer your question about the best practice, I think it would be to include some date/time value in the datamart itself. For instance, if Hubspot tracks when lead/contact data changes, that would be useful. However, I don't think this   data is present.

Reply:
Ah yeah, it would overwrite the field, like we want. There are also ways to add or remove from an array attribute like this, but I think we just want overwrite. Gory details here: https://www.braze.com/docs/api/objects_filters/user_attributes_object/

HOWEVER, while looking at that doc page, I found [documentation](https://www.braze.com/docs/user_guide/onboarding_with_braze/data_points/#data-points) talking about how every time we talk to braze to update attributes like this, it costs against our "data points" and thus costs money.   :-/

Even if Hubspot doesn't track when lead data changes, can the data mart definition in https://github.com/edx/warehouse-transforms/pull/1897 be changed to track that? Is there a way to update a `last_modified` field when the data mart sees a change? The default manifestation for our data marts is some special table yeah? So we do have a copy of the data sitting there, before we update the table?

I'm not super familiar with snowflake, so might need a pointer to an example or some help, but I could try to augment my data mart PR.

Reply:
Do you know how big this data will be? Are we talking hundreds of rows or millions? I assume it won't be too big. Also, how up to date does it need to be in Braze? In other words, how often will we need to run the Prefect flow? Both of those questions will help us determine how much we should care about the data point usage. I don't know about these points, but it will be useful to have some rough idea of the magnitude of this data flow.

Reply:
These are very practical questions. Let me talk to marketing and get back to you here.

Reply:
OK. Got word back from marketing. Looks like we're a drop in the bucket for now. And we have enough room in our api allocation to handle twice-daily updates (even when considering the next update to this task, to add aliases to unregistered email addresses).

So for now, we're not concerned with optimizing this. But we can always re-examine that if api capacity gets tight.

From that perspective, I think I'm happy with this task being as dumb as it is, on my side.

Reply:
Ok, cool. Thanks for checking.